File: src/api/modules/evaluation.js
import service from '@/api/request'
import { downloadFile } from '@/utils/download'

/**
 * 模型效果评估接口（已对接后端 FastAPI）。
 * 列表类读库；自动化指标/基准对比/场景验证为后端聚合分析结果。
 * 页面组件无需改动，函数签名与原 mock 版本保持一致。
 */

// 评估任务列表
export function getEvalList(params = {}) {
  return service.get('/evaluation/list', { params })
}

// 自动化指标（按模型类型）
export function getMetrics(modelType = 'ner') {
  return service.get('/evaluation/metrics', { params: { modelType } })
}

// 运行真实评估（payload: { modelId, datasetId, limit }）→ { evalTaskId, modelType }
export function runEvaluation(payload) {
  return service.post('/evaluation/run', payload)
}

// 轮询评估进度 / 结果
export function getEvaluationRun(evalTaskId) {
  return service.get(`/evaluation/run/${evalTaskId}`)
}

// 基准对比
export function getBenchmark() {
  return service.get('/evaluation/benchmark')
}

// 运行基准对比（微调模型 vs 基座模型，payload: { modelId, datasetId, limit }）
export function runBenchmark(payload) {
  return service.post('/evaluation/benchmark/run', payload)
}

// 轮询基准对比进度
export function getBenchmarkRun(evalTaskId) {
  return service.get(`/evaluation/benchmark/run/${evalTaskId}`)
}

// 真实业务场景验证
export function getSceneValidation() {
  return service.get('/evaluation/scene-validation')
}

// 运行场景验证（payload: { modelId, datasetIds, limit }）
export function runSceneValidation(payload) {
  return service.post('/evaluation/scene/run', payload)
}

// 轮询场景验证进度
export function getSceneRun(evalTaskId) {
  return service.get(`/evaluation/scene/run/${evalTaskId}`)
}

// 人工复核
export function getReviewSamples(params = {}) {
  return service.get('/evaluation/review-samples', { params })
}

export function getReviewSummary() {
  return service.get('/evaluation/review-summary')
}

// 复核抽样：真模型预测入复核队列（payload: { modelId, datasetId, count, reviewer }）
export function runReviewSampling(payload) {
  return service.post('/evaluation/review/sample', payload)
}

// 轮询复核抽样进度
export function getReviewRun(evalTaskId) {
  return service.get(`/evaluation/review/run/${evalTaskId}`)
}

// 错误案例分析
export function getErrorCases(params = {}) {
  return service.get('/evaluation/error-cases', { params })
}

// 评估报告
export function getReportList(params = {}) {
  return service.get('/evaluation/reports', { params })
}

// 生成评估报告（payload: { model, modelType, sections, format }）
export function generateReport(payload) {
  return service.post('/evaluation/reports', payload)
}

// 删除评估报告
export function deleteReport(reportId) {
  return service.delete(`/evaluation/reports/${reportId}`)
}

// 提交人工复核结果（results: [{ id, result }]）
export function submitReviewResults(results) {
  return service.post('/evaluation/review-results', { results })
}

// 导出评估报告（format: 'pdf' | 'excel'，浏览器另存为）
export function exportReport(reportId, format = 'pdf') {
  return downloadFile(`/evaluation/reports/${reportId}/export`, { params: { format } })
}

// 导出错误案例为 Excel（可按错误类型过滤）
export function exportErrorCases(errorType = '') {
  return downloadFile('/evaluation/error-cases/export', { params: { errorType } })
}

File: server/app/routers/evaluation.py
"""模型效果评估接口，路径与前端 src/api/modules/evaluation.js 对应。"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core import storage
from app.core.database import get_db
from app.core.response import ok, err, page
from app.crud import evaluation as crud
from app.deps import get_current_user
from app.models.user import User
from app.schemas.evaluation import (
    EvalTaskOut, ReviewSampleOut, ErrorCaseOut, EvalReportOut,
    ReportGenIn, ReviewSubmitIn, EvalRunIn, SceneRunIn, ReviewSampleIn,
)
from app.services import reporting
from app.services.eval_engine import runner as eval_runner

router = APIRouter(prefix="/evaluation", tags=["evaluation"])

_XLSX_MIME = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"


@router.get("/list")
def get_eval_list(
    page_no: int = Query(1, alias="page"),
    page_size: int = Query(10, alias="pageSize"),
    db: Session = Depends(get_db),
):
    items, total = crud.list_eval_tasks(db, page_no, page_size)
    return ok(page([EvalTaskOut.model_validate(x) for x in items], total, page_no, page_size))


@router.get("/metrics")
def get_metrics(modelType: str = "ner", db: Session = Depends(get_db)):
    return ok(crud.metrics(db, modelType))


@router.post("/run")
def run_evaluation(
    body: EvalRunIn,
    db: Session = Depends(get_db),
    current: User = Depends(get_current_user),
):
    """真跑评估：加载所选模型版本，在已发布测试集上推理算指标并落库。"""
    creator = current.real_name or current.username
    try:
        result = eval_runner.start(body.modelId, body.datasetId, body.limit, creator)
    except Exception as e:
        return err(f"启动评估失败：{e}", code=4001)
    return ok(result)


@router.get("/run/{eval_task_id}")
def get_evaluation_run(eval_task_id: int, db: Session = Depends(get_db)):
    """轮询评估进度 / 结果。"""
    status = eval_runner.status(db, eval_task_id)
    if not status:
        return err("评估任务不存在", code=4004)
    return ok(status)


@router.get("/benchmark")
def get_benchmark(db: Session = Depends(get_db)):
    return ok(crud.benchmark(db))


@router.post("/benchmark/run")
def run_benchmark(
    body: EvalRunIn,
    db: Session = Depends(get_db),
    current: User = Depends(get_current_user),
):
    """真跑基准对比：所选微调版本 vs 其基座模型，在同一已发布测试集上各评估一遍。"""
    creator = current.real_name or current.username
    try:
        result = eval_runner.start_benchmark(body.modelId, body.datasetId, body.limit, creator)
    except Exception as e:
        return err(f"启动基准对比失败：{e}", code=4001)
    return ok(result)


@router.get("/benchmark/run/{eval_task_id}")
def get_benchmark_run(eval_task_id: int, db: Session = Depends(get_db)):
    """轮询基准对比进度。"""
    status = eval_runner.status(db, eval_task_id)
    if not status:
        return err("对比任务不存在", code=4004)
    return ok(status)


@router.get("/scene-validation")
def get_scene_validation(db: Session = Depends(get_db)):
    return ok(crud.scene_validation(db))


@router.post("/scene/run")
def run_scene_validation(
    body: SceneRunIn,
    db: Session = Depends(get_db),
    current: User = Depends(get_current_user),
):
    """真跑业务场景验证：所选模型在多个已发布测试集上各评估一遍，每个测试集记为一个场景。"""
    creator = current.real_name or current.username
    try:
        result = eval_runner.start_scene(body.modelId, body.datasetIds, body.limit, creator)
    except Exception as e:
        return err(f"启动场景验证失败：{e}", code=4001)
    return ok(result)


@router.get("/scene/run/{eval_task_id}")
def get_scene_run(eval_task_id: int, db: Session = Depends(get_db)):
    """轮询场景验证进度。"""
    status = eval_runner.status(db, eval_task_id)
    if not status:
        return err("场景验证任务不存在", code=4004)
    return ok(status)


@router.get("/review-samples")
def get_review_samples(
    page_no: int = Query(1, alias="page"),
    page_size: int = Query(10, alias="pageSize"),
    db: Session = Depends(get_db),
):
    items, total = crud.list_review_samples(db, page_no, page_size)
    return ok(page([ReviewSampleOut.model_validate(x) for x in items], total, page_no, page_size))


@router.get("/review-summary")
def get_review_summary(db: Session = Depends(get_db)):
    return ok(crud.review_summary(db))


@router.post("/review/sample")
def run_review_sampling(
    body: ReviewSampleIn,
    db: Session = Depends(get_db),
    current: User = Depends(get_current_user),
):
    """真模型对所选测试集抽样预测，入人工复核队列（待复核）。"""
    creator = current.real_name or current.username
    reviewer = body.reviewer or creator
    try:
        result = eval_runner.start_review(body.modelId, body.datasetId, body.count, reviewer, creator)
    except Exception as e:
        return err(f"启动复核抽样失败：{e}", code=4001)
    return ok(result)


@router.get("/review/run/{eval_task_id}")
def get_review_run(eval_task_id: int, db: Session = Depends(get_db)):
    """轮询复核抽样进度。"""
    status = eval_runner.status(db, eval_task_id)
    if not status:
        return err("复核抽样任务不存在", code=4004)
    return ok(status)


@router.get("/error-cases")
def get_error_cases(
    errorType: str = "",
    page_no: int = Query(1, alias="page"),
    page_size: int = Query(10, alias="pageSize"),
    db: Session = Depends(get_db),
):
    items, total, dist = crud.list_error_cases(db, errorType, page_no, page_size)
    result = page([ErrorCaseOut.model_validate(x) for x in items], total, page_no, page_size)
    result["dist"] = dist
    return ok(result)


@router.get("/reports")
def get_report_list(
    page_no: int = Query(1, alias="page"),
    page_size: int = Query(10, alias="pageSize"),
    db: Session = Depends(get_db),
):
    items, total = crud.list_reports(db, page_no, page_size)
    return ok(page([EvalReportOut.model_validate(x) for x in items], total, page_no, page_size))


@router.post("/reports")
def generate_report(
    body: ReportGenIn,
    db: Session = Depends(get_db),
    current: User = Depends(get_current_user),
):
    creator = current.real_name or current.username
    rep = crud.create_report(db, model=body.model, creator=creator)
    return ok(EvalReportOut.model_validate(rep))


@router.delete("/reports/{report_id}")
def delete_report(report_id: int, db: Session = Depends(get_db)):
    """删除一份评估报告。"""
    if not crud.delete_report(db, report_id):
        return err("报告不存在", code=4004)
    return ok({"deleted": report_id})


@router.post("/review-results")
def submit_review_results(body: ReviewSubmitIn, db: Session = Depends(get_db)):
    updated = crud.submit_review_results(db, [r.model_dump() for r in body.results])
    return ok({"updated": updated})


@router.get("/error-cases/export")
def export_error_cases(errorType: str = "", db: Session = Depends(get_db)):
    """导出错误案例为 Excel 文件。"""
    rows = crud.all_error_cases(db, errorType)
    data = reporting.error_cases_excel(rows)
    name = f"错误案例分析{('-' + errorType) if errorType else ''}.xlsx"
    _stored, abspath, _size = storage.save_bytes("reports", name, data)
    return storage.file_response(abspath, name, media_type=_XLSX_MIME)


@router.get("/reports/{report_id}/export")
def export_report(report_id: int, format: str = "pdf", db: Session = Depends(get_db)):
    """导出评估报告：format=pdf 生成 PDF，format=excel 生成 Excel。"""
    rep = crud.get_report(db, report_id)
    if not rep:
        return err("报告不存在", code=4004)
    if format == "excel":
        data = reporting.report_excel(rep)
        name, mime = f"{rep.name}.xlsx", _XLSX_MIME
    else:
        data = reporting.report_pdf(rep)
        name, mime = f"{rep.name}.pdf", "application/pdf"
    _stored, abspath, _size = storage.save_bytes("reports", name, data)
    return storage.file_response(abspath, name, media_type=mime)
